File: python/zmq/clonecli1.py
# clone client model one
import random, time, zmq
from kvsimple import KVMsg

def main():
	ctx = zmq.Context()
	updates = ctx.socket(zmq.SUB)
	updates.linger = 0
	updates.setsockopt(zmq.SUBSCRIBE, '')
	updates.connect('tcp://localhost:5556')
	kvmap = {}
	sequence = 0
	while True:
		try:
			kvmsg = KVMsg.recv(updates)
		except KeyboardInterrupt:
			break  # interrupted
		kvmsg.store(kvmap)
		sequence += 1
	print 'Interrupted\n%d messages in' % sequence

if __name__ == '__main__':
	main()

File: zmq3/clonecli1.cpp
// clone client model one (see figure 58 from zguide)
#include <czmq.h>
#include "kvsimple.h"

int main(int argc, char * argv[])
{
	zctx_t * ctx = zctx_new();
	void * updates = zsocket_new(ctx, ZMQ_SUB);
	zsocket_set_subscribe(updates, "");
	zsocket_connect(updates, "tcp://localhost:5556");

	zhash_t * kvmap = zhash_new();
	int64_t sequence = 0;

	while (!zctx_interrupted)
	{
		kvmsg_t * kvmsg = kvmsg_recv(updates);
		if (!kvmsg)
			break;
		kvmsg_store(&kvmsg, kvmap);
		sequence++;
	}

	printf(" Interrupted\n%d messages in\n", (int)sequence);
	zhash_destroy(&kvmap);
	zctx_destroy(&ctx);

	return 0;
}
